### PR TITLE
use ocaml 4.12.0

### PR DESCRIPTION
**Pull request by idkjs**
This PR gets the frontend working again. We should be able to run `make db` and `yarn start` and be good to go after setting up the frontend and the backend.

It also updates the compiler to 4.12.0 so we can use that if you have it installed globally and dont want to wait for a compiler to build.

It uses `bs-platform` for the frontend and has its own `counter.atd` atdgen set up.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -12,6 +12,14 @@ help:
 	@grep -E '^[a-zA-Z_-]+:.*?## .*$$' $(MAKEFILE_LIST) | sort | awk 'BEGIN {FS = ":.*?## "}; {printf "  \033[36m%-15s\033[0m %s\n", $$1, $$2}';
 	@echo "";
 
+.PHONY: db
+db: ## Start the database with dune exec
+	dune exec counter
+
+.PHONY: front
+front: ## Start the front with yarn start
+	yarn start
```
Comment (davesnx) [inline Makefile:21]: The makefile shoudn't call yarn

File: Makefile [hunk 2]
```diff
@@ -26,7 +34,7 @@ install: ## Install the dependencies
 
 .PHONY: init
 init: ## Create a local opam switch and install deps
-	opam switch create . 4.10.0 --deps-only --locked
+	# opam switch create . 4.12.0 --deps-only --locked
```
Comment (davesnx) [inline Makefile:37]: Why comment this?

File: frontend/atd/Counter_bs.ml [hunk 3]
```diff
@@ -0,0 +1,153 @@
+(* Auto-generated from "Counter.atd" *)
```
Comment (davesnx) [inline frontend/atd/Counter_bs.ml:1]: Do you think we should have this inside git?

File: frontend/components/Api.re [hunk 4]
```diff
@@ -0,0 +1,107 @@
+let baseUrl = "http://127.0.0.1:3003/";
```
Comment (davesnx) [inline frontend/components/Api.re:1]: We should keep Api outside components I believe.

File: frontend/components/App.re [hunk 5]
```diff
@@ -1,30 +1,38 @@
 open UI;
 
-module H1 = [%styled.h1 {|
+module H1 = [%styled.h1
+  {|
```
Comment (davesnx) [inline frontend/components/App.re:4]: Please disable refmt

File: package.json [hunk 6]
```diff
@@ -7,28 +7,33 @@
...
   "dependencies": {
-    "@davesnx/styled-ppx": "^1.5.5",
+    "@ahrefs/bs-atdgen-codec-runtime": "^2.0.1",
+    "@davesnx/styled-ppx": "1.5.6",
     "@rescript/react": "^0.10.3",
-    "bs-css": "^14.0.2",
     "bs-css-emotion": "^3.0.0",
     "bs-fetch": "^0.6.2",
     "bs-webapi": "^0.19.1",
+    "isomorphic-fetch": "^3.0.0",
```
Comment (davesnx) [inline package.json:28]: We don't use a isomophic fetch anywhere in this repo, why do we have this dep?

File: package.json [hunk 7]
```diff
@@ -7,28 +7,33 @@
...
+    "bundle:watch": "yarn bundle --watch --color=true",
     "dev-server": "PORT=5000 live-server ./www --host=localhost",
-    "bs:build": "NINJA_ANSI_FORCED=1 esy x bsb -make-world",
-    "bs:watch": "watchexec -w frontend --exts re,rei,ml,mli,atd,json -r -c 'npm run bs:build'",
-    "bs:clean": "esy x bsb -clean-world",
-    "start": "npm run clean; concurrently --names re,dev,js --kill-others 'npm run bs:watch' 'npm run dev-server' 'npm run bundle:watch'",
-    "bundle:prod": "npm run bundle --minify --define:process.env.NODE_ENV='\"production\"'",
-    "clean": "rm -rf www/js; npm run bs:clean",
+    "bs:build": "NINJA_ANSI_FORCED=1 bsb -make-world",
+    "bs:watch": "NINJA_ANSI_FORCED=1 bsb -make-world -w",
```
Comment (davesnx) [inline package.json:14]: Lets keep using bsb from melange and keep the watchexec since watch doesn't work well in melange

File: Makefile (hunk 2 above)
Comment (idkjs) [reply Makefile:37]: it was commented so I could run `make init` without creating a switch. Could have just run `make install`. Fixed it.

File: frontend/components/Api.re (hunk 4 above)
Comment (idkjs) [reply frontend/components/Api.re:1]: fixed in c42c2f534d0c26684404f0db933fdad5934f49b9

File: package.json (hunk 6 above)
Comment (idkjs) [reply package.json:28]: removed dep in c802c7388fdd112cdab1f5c7cb4b819eebe6568c

File: frontend/atd/Counter_bs.ml (hunk 3 above)
Comment (idkjs) [reply frontend/atd/Counter_bs.ml:1]: I have added atdgen output to `.gitignore`

File: package.json (hunk 7 above)
Comment (idkjs) [reply package.json:14]: How do we use bsb from melange and not get a conflict? Going to try to see if the new folder structure makes a difference by running it. I dont think it works for a few different reasons though so it probably wont work. Which is why I took it out.

This is what you get from trying to run `yarn start` after building with `esy bsb -make-world`:
```sh
[dev]  > node_modules/bs-fetch/src/Fetch.bs.js:4:24: error: Could not resolve "bs-platform/lib/es6/stdlib.js" (mark it as external to exclude it from the bundle)
[dev]     4 │ import * as Stdlib from "bs-platform/lib/es6/stdlib.js";
[dev]       ╵                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[dev] 
[dev]  > node_modules/@glennsl/bs-json/src/Json_decode.bs.js:6:24: error: Could not resolve "bs-platform/lib/es6/stdlib.js" (mark it as external to exclude it from the bundle)
[dev]     6 │ import * as Stdlib from "bs-platform/lib/es6/stdlib.js";
```

I tried this too but maybe needs to be tweaked. Doesnt work.

```sh
.PHONY: watch-front
watch-front: ## Watch changes on source-files and re-run the server
	@watchexec -w frontend \
		--exts re,rei,ml,mli,atd,json -r -c \
		"clear; $(MAKE) run-debug"
```
looks like its still reading the backend dir
```sh
opam exec -- dune exec counter -- --debug
File "backend/dune", line 8, characters 27-37:
8 |   (pps ppx_deriving_yojson ppx_yojson ppx_rapper)))
                               ^^^^^^^^^^
Error: Library "ppx_yojson" not found.
Hint: try:
  dune external-lib-deps --missing counter
make[1]: *** [run-debug] Error 1
^Cmake: *** [watch-front] Interrupt: 2

~/Github/CAAS/backend-only main !2 ?2
```

**Pull request merged**